File: gsr/git-recency-search.ts
import { exec } from 'child_process';
import { promisify } from 'util';

const execAsync = promisify(exec);

interface SearchResult {
  filename: string;
  commit: string;
  match: string;
}

async function gitSearch(searchString: string, maxResults: number, contextLines: number): Promise<SearchResult[]> {
  const results: SearchResult[] = [];
  let commitHash = '';

  while (results.length < maxResults) {
    // Get the next commit
    const { stdout: commit } = await execAsync(`git log -n1 --format=%H ${commitHash}`);
    if (!commit.trim()) break; // No more commits

    // Get the diff for this commit
    const { stdout: diff } = await execAsync(`git show ${commit.trim()} -U${contextLines}`);

    // Search for the string in the diff
    const regex = new RegExp(`^(@@.*\\n)?(.*${searchString}.*(?:\\n.+){0,${contextLines}})`, 'gim');
    const matches = diff.matchAll(regex);

    for (const match of Array.from(matches)) {
      const fileMatch = diff.slice(0, match.index).match(/^--- a\/(.*)/m);
      if (fileMatch) {
        const filename = fileMatch[1];
        results.push({
          filename,
          commit: commit.trim(),
          match: match[0],
        });

        if (results.length >= maxResults) break;
      }
    }

    commitHash = `${commit.trim()}^`;
  }

  return results;
}

function parseArgs(): { searchString: string; maxResults: number; contextLines: number } {
  const args = process.argv.slice(2);
  let searchString = '';
  let maxResults = 10;
  let contextLines = 3;

  for (let i = 0; i < args.length; i++) {
    switch (args[i]) {
      case '-m':
      case '--max-results':
        maxResults = parseInt(args[++i], 10);
        break;
      case '-c':
      case '--context-lines':
        contextLines = parseInt(args[++i], 10);
        break;
      default:
        if (!searchString) searchString = args[i];
        break;
    }
  }

  if (!searchString) {
    console.error('Error: Search string is required.');
    process.exit(1);
  }

  return { searchString, maxResults, contextLines };
}

async function main() {
  const { searchString, maxResults, contextLines } = parseArgs();

  try {
    const results = await gitSearch(searchString, maxResults, contextLines);

    results.forEach((result, index) => {
      console.log(`File: ${result.filename}`);
      console.log(`Commit: ${result.commit}`);
      console.log(result.match);
      if (index < results.length - 1) {
        console.log('-'.repeat(80));
      }
    });
  } catch (error) {
    console.error('An error occurred:', error);
  }
}

main();

File: gsr/git-recency-search.py
import subprocess
import re
import argparse

def git_search(search_string, max_results, context_lines):
    commit_hash = ''
    results = []
    
    while len(results) < max_results:
        # Git command to get the diff of the next commit
        git_command = f"git log -n1 --format=%H {commit_hash}".split()
        commit = subprocess.check_output(git_command).decode().strip()
        
        if not commit:
            break  # No more commits
        
        # Get the diff for this commit
        diff_command = f"git show {commit} -U{context_lines}".split()
        diff = subprocess.check_output(diff_command).decode()
        
        # Search for the string in the diff
        matches = re.finditer(f"^(@@.*\n)?([^\n]*{re.escape(search_string)}[^\n]*(\n[^\n]+){{0,{context_lines}}})", diff, re.MULTILINE)
        
        for match in matches:
            file_match = re.search(r"^--- a/(.*)", match.string[:match.start()], re.MULTILINE)
            if file_match:
                filename = file_match.group(1)
                result = f"File: {filename}\nCommit: {commit}\n{match.group()}\n"
                results.append(result)
                
                if len(results) >= max_results:
                    break
        
        commit_hash = f"{commit}^"
    
    return results

def main():
    parser = argparse.ArgumentParser(description="Search for a string in Git commits")
    parser.add_argument("search_string", help="String to search for")
    parser.add_argument("--max-results", type=int, default=10, help="Maximum number of results to return")
    parser.add_argument("--context-lines", type=int, default=3, help="Number of context lines to show")
    
    args = parser.parse_args()
    
    results = git_search(args.search_string, args.max_results, args.context_lines)
    
    for result in results:
        print(result)
        print("-" * 80)

if __name__ == "__main__":
    main()
